### ISSUE TITLE
Allow sharing of vendor dir (CI server usecase)

### ISSUE THREAD

**Issue by gggeek** [composer/composer#4174]
I think this has been requested already multiple times, in slightly different forms: #1940 #1599 #1758 #664.
Since all of them are closed, I'm opening a new issue :-)

To define the precise scope of it and avoid shooting for pie-in-the-sky or completely-against-composer-principles:
- not sharing the vendor dir between projects is the greatest thing ever
- except when you run a build server which runs the same test job hundreds of times with different test configurations or from different branches of the sources
- in that case, disk space used by vendor folders can accumulate quickly, as well as the time needed to run all the 'composer install' command
- running in parallel many tests from one codebase is hard, so just sharing the whole project directory does not fly (because there might be configs which the project expect to be in a specific file and can not be overridden easily per-test, or hardcoded cache folders or logs or folders where artifacts are generated, etc... making sure many tests can run in parallel in isolation from the same codebase folder is just asking too much from / trusting too much the devs)

So

To make the platform leaner, I tried sharing the 'vendor' folder via automation:
- using symlinks
- basing the sharing on md5 of composer.lock

It mostly works - and I do not advocate pushing this into composer as core feature.

I just found out that:
1. after moving a 'vendor' folder away and symlinking it back, I have to run dumpautoload again, which is fine
2. the $baseDir variable which gets generated by composer for its autoloaders gets hardcoded to the directory used to run the original 'dumpautoload' command. This is a blocker.

The solution I came up with: patch by hand the generated autoload files - but it is hackish, and brittle.

Is there a better way to achieve this?
How could I inject the $baseDir value into the generated autoloader, so that each project using the same shared vendor can set it to a different value at runtime?

**Comment by staabm** [comment]
This doesnt sound like a problem which need to be solved at composer lvl.

CI systems like e.g. Travis support sharing of dependencies between builds:
http://docs.travis-ci.com/user/caching/#Caching-directories-%28Bundler%2C-dependencies%29

**Comment by gggeek** [comment]
@staabm I agree that Composer does not need to take fully the responsibility for that.
That's why I am asking for just a small alteration so far - to make it easier to inject one variable which at the moment is hardcoded ;-)
Otoh there are other CI systems than Travis, and I am pretty sure not all of them have good support it natively.

**Comment by Seldaek** [comment]
Is this still needed/relevant @gggeek?

**Comment by gggeek** [comment]
Well, we managed to get new VMs with more disk allocated, but in the meantime the number of tests we run has also increased... Also, we found no better solution/workaround than what is described above.
So yes, I'd say it is still relevant

**Comment by curry684** [comment]
I'm not really sure we can fix this. The root issue is that we really don't lock anything at all to any folder, but PHP is. All autoload files start with (a form of):
```php
$vendorDir = dirname(dirname(__FILE__));
$baseDir = dirname($vendorDir);
```
`__FILE__` is a magical constant representing the full path to the currently executing file (it should actually use `__DIR__` in this case but that's a different issue). The problem is that PHP is blissfully unaware there of the existence of symlinks, and normalizes that path. Which is a good thing because it makes opcache and mechanisms like `require_once` work correctly. But it also means that essentially we have no way of detecting otherwise.

You can verify this easily:
```sh
# mkdir example
# echo "<?php var_dump(__FILE__);" > example/test.php
# ln -s example symlink
# php symlink/test.php
string(28) "/home/<redacted>/example/test.php"
```
So the only thing we could do about this is some kind of magic constant setting, like:
```php
$vendorDir = defined('PROJECT_VENDOR') ? PROJECT_VENDOR : dirname(dirname(__FILE__));
$baseDir = dirname($vendorDir);
```
Seems out of proportion to incur these extra CPU cycles on the whole world though on every execution.

**Comment by Seldaek** [comment]
We could maybe do it at autoload-dump time, but indeed not in autoload files themselves. It wouldn't be flexible, would probably make all paths absolute etc, but it might work. Not sure if really worth it though. As @gggeek does not need it anymore I would let this sit here a while and revisit later if there is still a need.

**Comment by gggeek** [comment]
ok

**Comment by Seldaek** [comment]
Closing because well.. most likely not gonna happen sorry
